operator/controllers/instance_controller.go

```diff
@@ -0,0 +1,461 @@
+
+const instanceReconcileTimeout = time.Second * 10
+
+// Reconcile reconciles an Instance resource.
+func (r *InstanceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), instanceReconcileTimeout)
+	defer cancel()
+
+	log := r.Log.WithValues("instance", req.NamespacedName)
+
+	instance := &servicebrokerv1alpha1.Instance{}
+	if err := r.Get(ctx, req.NamespacedName, instance); err != nil {
+		if errors.IsNotFound(err) {
+			instanceName := req.NamespacedName.Name
+			helmInstanceName := fmt.Sprintf("metabroker-%s", instanceName)
+			namespace := req.NamespacedName.Namespace
+			podName := fmt.Sprintf("metabroker-%s-deprovision", instanceName)
+			return r.deprovisioningPod(ctx, helmInstanceName, podName, namespace)
+		}
+		return ctrl.Result{}, err
+	}
+
+	planNamespacedName := req.NamespacedName
+	planNamespacedName.Name = instance.Spec.Plan
+	plan := &servicebrokerv1alpha1.Plan{}
+	if err := r.Get(ctx, planNamespacedName, plan); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	instanceNeedsUpdate := false
+
+	if len(instance.OwnerReferences) == 0 {
+		if err := r.setOwnership(ctx, instance, plan); err != nil {
+			return ctrl.Result{}, nil
+		}
+		instanceNeedsUpdate = true
+	}
+
+	if instance.Spec.ID == "" {
+		id := uuid.Must(uuid.NewUUID()) // UUID v1
+		instance.Spec.ID = id.String()
+		instanceNeedsUpdate = true
+	}
+
+	if instance.Spec.ValidateValues == nil {
+		validateValues := true
+		instance.Spec.ValidateValues = &validateValues
+		instanceNeedsUpdate = true
+	}
+
+	if instanceNeedsUpdate {
+		if err := r.Update(ctx, instance); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// TODO: update the status with validated so the controller doesn't keep performing the
+	// validation on requeues.
+	if *instance.Spec.ValidateValues {
+		schemaJSON, err := yaml.YAMLToJSON([]byte(plan.Spec.Provisioning.Values.Schema))
+		if err != nil {
+			log.Error(err, "failed to validate values")
+			// TODO: update status with invalid values.
+			return ctrl.Result{}, nil
+		}
+		schema := gojsonschema.NewBytesLoader(schemaJSON)
+		valuesJSON, err := yaml.YAMLToJSON([]byte(instance.Spec.Values))
+		if err != nil {
+			log.Error(err, "failed to validate values")
+			// TODO: update status with invalid values.
+			return ctrl.Result{}, nil
+		}
+		values := gojsonschema.NewBytesLoader(valuesJSON)
+		result, err := gojsonschema.Validate(schema, values)
+		if err != nil {
+			log.Error(err, "failed to validate values")
+			// TODO: update status with invalid values.
+			return ctrl.Result{}, nil
+		}
+		if !result.Valid() {
+			log.Error(err, "failed to validate values")
+			// TODO: update status with invalid values. Including specific errors returned from
+			// result.Errors().
+			return ctrl.Result{}, nil
+		}
+	}
+
+	valuesSecretName := fmt.Sprintf("metabroker-%s-values", instance.Name)
```

Comment:
Don't really like having the names be coming from `printf` calls scattered all over the place (because it's easy to get out of sync); would be nice if this can be a help method on `instance`.

Of course, it would be even better if it's a custom type so we can be sure it's used correctly, but that's probably overkill?

That, and it would be a lot better to be passing around `types.NamedspacedName`s rather than strings.

Reply:
While reading your comment, I don't think it's overkill to have a type for this. I'll see if that indeed makes more sense in the code.

Reply:
I think 1b1b271 makes more sense.